clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -155,7 +155,10 @@ public void run() {
             this.accumulator.abortIncompleteBatches();
         }
         try {
-            this.client.close();
+            if (!forceClose)
+                this.client.closeGracefully();
```

Comment:
@ijuma Oops! You are absolutely right. The current fix doesn't help here. I had forgotten that the issue I intended to fix when I raised the JIRA was to ensure that the data does indeed get flushed before close returns. Needs some more work. I will close this PR and reopen once I have a full fix. Thank you!